trame_server/client.py

```diff
@@ -0,0 +1,302 @@
+                break
+
+    async def auth(self, **kwargs):
+        key = WsLinkSession.AUTH_ID
+        resp = self.loop.create_future()
+        msg = dict(
+            wslink="1.0",
+            id=key,
+            method="wslink.hello",
+            args=[kwargs],
+            kwargs={},
+        )
+        self.in_flight_rpc[key] = resp
+        await self.ws.send_str(json.dumps(msg))
+        return resp
+
+    async def call(self, method, args=None, kwargs=None):
+        self.msg_count += 1
+        key = f"rpc:{self.client_id}:{self.msg_count}"
+        resp = self.loop.create_future()
+        self.in_flight_rpc[key] = resp
+        if args is None:
+            args = []
+        if kwargs is None:
+            kwargs = {}
+
+        await self.ws.send_str(
+            json.dumps(
+                dict(
+                    wslink="1.0",
+                    id=key,
+                    method=method,
+                    args=args,
+                    kwargs=kwargs,
+                )
+            )
+        )
+
+        return resp
+
+    def register_subscription(self, topic, callback):
+        if topic not in self.subscriptions:
+            self.subscriptions[topic] = [callback]
+        else:
+            self.subscriptions[topic].append(callback)
+
+    def unregister_subscription(self, topic, callback):
+        callbacks = self.subscriptions.get(topic, [])
+        if callback in callbacks:
+            callbacks.remove(callback)
+
+        if len(callbacks) == 0 and topic in self.subscriptions:
+            self.subscriptions.pop(topic)
+
+    def clear_subscriptions(self):
+        topics = list(self.subscriptions.keys())
+        for topic in topics:
+            self.subscriptions.pop(topic)
+
+
+class Client:
+    """
+    Client implementation for driving a remote trame server with its shared state and trigger method calls in plain python.
+    """
+
+    def __init__(self, hot_reload=False):
+        # Network
+        self._connected = False
+        self._session = None
+
+        # fake server
+        self.hot_reload = hot_reload
+        self._change_callbacks = {}
+
+        # trame state
+        self._state = State(self)
+
+    async def connect(self, url, **kwargs):
+        async with aiohttp.ClientSession() as session:
+            async with session.ws_connect(url) as ws:
+                self._session = WsLinkSession(ws)
+                self._state.server_ready()
+                self._session.register_subscription(
+                    "trame.state.topic", self._on_state_update
+                )
+                self._connected = True
+                task = asynchronous.create_task(self._session.listen())
+                await self._session.auth(**kwargs)
+                await task
+
+        self._session.clear_subscriptions()
+        self._session = None
+        self._connected = False
+
+    async def diconnect(self):
+        if self._ws:
+            await self._ws.close()
```

Comment:
Looks like `self._ws` doesn't get set anywhere.

Reply:
Good catch

Reply:
`self.ws` isn't set anywhere on the `Client` either, FYI.